geo/src/algorithm/sweep/mod.rs

```diff
@@ -1,29 +1,364 @@
+///
+/// # Event Relationships
+///
+/// Each line segment generates two events: an `INSERT` event at its left end and
+/// a `DELETE` event at its right end. These paired events are correlated using the
+/// segment index stored in the `SweepLineInterval`.
+///
+/// Rather than explicit pointers or references, events are stored in two separate vecs
+/// (`insert_events` and `delete_events`) and correlated through their shared segment index.
+/// A lookup array maps segment indices to DELETE x-coordinates for access during
+/// the sweep process.
+///
+/// # Implementation Note: Why No Binary Search Tree or Priority Queue?
+///
+/// The Bentley-Ottmann algorithm as described typically uses two tracking data structures:
+///
+/// 1. A **priority queue** to process events in coordinate order
+/// 2. A **balanced binary search tree** to track active segments sorted by their `y`-coordinate position
+///
+/// This implementation avoids both:
+///
+/// - Instead of a priority queue, the **event vecs are sorted once** and process events in order.
+///   This has the same asymptotic complexity.
+///
+/// - Instead of a binary search tree, we use **two separate vecs** for INSERT and DELETE events.
+///   When processing an `INSERT` event, we find its corresponding DELETE event by segment index
+///   and check all INSERT events that start before that DELETE x-coordinate.
+///
+/// By maintaining separate vecs for INSERT and DELETE events and using the segment index
+/// for correlation, we can determine which segments are active at any point
+/// _without_ maintaining an "active segment" tree or using a HashMap.
+#[derive(Debug, Clone)]
+struct SweepLineEvent<T: GeoFloat> {
+    x_value: T,
+    event_type: EventType,
+    interval: SweepLineInterval<T>,
+}
+
+impl<T: GeoFloat> SweepLineEvent<T> {
+    fn new(x: T, event_type: EventType, interval: SweepLineInterval<T>) -> Self {
+        Self {
+            x_value: x,
+            event_type,
+            interval,
+        }
+    }
+}
+
+impl<T: GeoFloat> PartialEq for SweepLineEvent<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.x_value.total_cmp(&other.x_value) == Ordering::Equal
+            && self.event_type == other.event_type
+    }
+}
+
+impl<T: GeoFloat> Eq for SweepLineEvent<T> {}
+
+impl<T: GeoFloat> PartialOrd for SweepLineEvent<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T: GeoFloat> Ord for SweepLineEvent<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Use total_cmp for more robust floating point comparison
+        // This is crucial for grid patterns where many x-values are exactly equal
+        match self.x_value.total_cmp(&other.x_value) {
+            Ordering::Equal => {
+                // When x values are equal, prioritize INSERT events before DELETE events
+                // This ensures proper handling of overlapping segments
```

Comment:
I can't see where we ever have a mixed collection of INSERT and DELETE events - so it seems like this detail of the sorting implementation has no effect. 

Unless I'm missing something we could delete the `event_type` field altogether - we already infer the event_type by if it lives in the `index.insert_events` field or the `index.delete_events` field.

Reply:
Yep, we can absolutely get rid of this. Nice.